src/main/java/com/google/cloud/tools/app/cloudsdk/CloudSdkAppEngineFactory.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * Copyright 2016 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.app.cloudsdk;
+
+import com.google.cloud.tools.app.api.AppEngineFactory;
+import com.google.cloud.tools.app.api.deploy.AppEngineDeployment;
+import com.google.cloud.tools.app.api.deploy.AppEngineFlexibleStaging;
+import com.google.cloud.tools.app.api.deploy.AppEngineStandardStaging;
+import com.google.cloud.tools.app.api.devserver.AppEngineDevServer;
+import com.google.cloud.tools.app.api.genconfig.GenConfigUtility;
+import com.google.cloud.tools.app.api.instances.AppEngineInstances;
+import com.google.cloud.tools.app.api.logs.AppEngineLogs;
+import com.google.cloud.tools.app.api.services.AppEngineServices;
+import com.google.cloud.tools.app.api.versions.AppEngineVersions;
+import com.google.cloud.tools.app.cloudsdk.internal.CloudSdkAppEngineDeployment;
+import com.google.cloud.tools.app.cloudsdk.internal.CloudSdkAppEngineDevServer;
+import com.google.cloud.tools.app.cloudsdk.internal.CloudSdkAppEngineFlexibleStaging;
+import com.google.cloud.tools.app.cloudsdk.internal.CloudSdkAppEngineGenConfig;
+import com.google.cloud.tools.app.cloudsdk.internal.CloudSdkAppEngineInstances;
+import com.google.cloud.tools.app.cloudsdk.internal.CloudSdkAppEngineLogs;
+import com.google.cloud.tools.app.cloudsdk.internal.CloudSdkAppEngineServices;
+import com.google.cloud.tools.app.cloudsdk.internal.CloudSdkAppEngineStandardStaging;
+import com.google.cloud.tools.app.cloudsdk.internal.CloudSdkAppEngineVersions;
+import com.google.cloud.tools.app.cloudsdk.internal.sdk.CloudSdk;
+import com.google.cloud.tools.app.cloudsdk.process.ProcessExitListener;
+import com.google.cloud.tools.app.cloudsdk.process.ProcessOutputLineListener;
+import com.google.cloud.tools.app.cloudsdk.process.ProcessStartListener;
+
+import java.io.File;
+
+/**
+ * A factory for creating Cloud SDK-based implementations of the App Engine services.
+ */
+public class CloudSdkAppEngineFactory implements AppEngineFactory {
+
+  private CloudSdk cloudSdk;
+
+  private CloudSdkAppEngineFactory(CloudSdk cloudSdk) {
+    this.cloudSdk = cloudSdk;
+  }
+
+  @Override
+  public AppEngineDeployment deployment() {
```

Comment:
'factory' methods  should be named in such a way as it's clear that it's creating something.  I also think the naming of the commands is a bit weird.  They sound like things rather than actions.

But more broadly, I do agree with @briandealwis and @loosebazooka that we may be struggling with getting the right abstractions and names here.